File: mac/FluentCore/Sources/FluentCore/GeminiLiveClient.swift
import Foundation

#if !canImport(FoundationNetworking)
/// Streams microphone audio to `gemini-3.5-transcribe-live` over the Live API WebSocket while the
/// user is still talking, so the transcript is (almost) ready the moment they stop. A port of
/// Android's `GeminiLiveClient`: same setup message, manual turn boundaries (the user says exactly
/// when dictation starts and stops), same finalise-then-await flow.
///
/// Audio sent before the socket is ready is queued and flushed on `setupComplete`. If the socket
/// never gets ready, or fails, `finish()` returns nil and the caller falls back to the batch
/// request with the full recording, so a turn is never lost.
public final class GeminiLiveClient: @unchecked Sendable {
    public let url: URL
    private let apiKey: String
    private let mode: TranscriptionMode
    private let languageCodes: [String]
    private let vocabulary: [String]
    private let session: URLSession

    private let lock = NSLock()
    private var task: URLSessionWebSocketTask?
    private var ready = false
    private var closed = false
    private var activityStarted = false
    private var queued: [Data] = []
    private var finals = ""
    private var turnDone = false
    /// True only when the server said the turn is complete: a stall, error or dropped socket leaves
    /// it false, and the caller then uses the batch fallback rather than a partial transcript.
    private var completedNormally = false
    private var waiter: CheckedContinuation<String?, Never>?
    public private(set) var failure: TranscriptionError?
    /// Latest interim (not yet final) text, for display.
    public var onInterim: (@Sendable (String) -> Void)?

    public init(apiKey: String, mode: TranscriptionMode, languageCodes: [String], vocabulary: [String],
                url: URL = Constants.liveURL, session: URLSession = .shared) {
        self.apiKey = apiKey
        self.mode = mode
        self.languageCodes = languageCodes
        self.vocabulary = vocabulary
        self.url = url
        self.session = session
    }

    public var isReady: Bool { lock.withLock { ready && !closed } }

    /// Opens the socket and sends the setup message. Returns immediately; audio can be sent at once.
    public func connect() {
        var comps = URLComponents(url: url, resolvingAgainstBaseURL: false)!
        comps.queryItems = (comps.queryItems ?? []) + [URLQueryItem(name: "key", value: apiKey)]
        var request = URLRequest(url: comps.url!)
        request.timeoutInterval = Constants.liveConnectTimeout
        let task = session.webSocketTask(with: request)
        lock.withLock { self.task = task }
        task.resume()
        send(Self.json(setupMessage()))
        receive()
        // Give up on the socket if it is not ready in time; the batch fallback takes over.
        DispatchQueue.global().asyncAfter(deadline: .now() + Constants.liveConnectTimeout) { [weak self] in
            guard let self else { return }
            if !self.lock.withLock({ self.ready }) { self.fail(.connectionLost) }
        }
    }

    /// Streams 16 kHz mono PCM16 samples. Safe to call from the audio thread.
    public func send(pcm: [Int16]) {
        guard !pcm.isEmpty else { return }
        let data = pcm.withUnsafeBufferPointer { Data(buffer: $0) }   // little-endian on every Mac
        let flushNow: Bool = lock.withLock {
            if closed { return false }
            if !ready { queued.append(data); return false }
            return true
        }
        if flushNow { send(Self.json(audioMessage(data))) }
    }

    /// Ends the utterance and waits for the final transcript. Nil means "use the batch fallback".
    public func finish(timeout: TimeInterval = Constants.liveFinalizeTimeout) async -> String? {
        let (isReady, started, alreadyDone): (Bool, Bool, Bool) = lock.withLock {
            (ready && !closed, activityStarted, turnDone)
        }
        if alreadyDone { close(); return normalResult() }
        guard isReady else { close(); return nil }
        if started { send(Self.json(["realtimeInput": ["activityEnd": [String: Any]()]])) }
        send(Self.json(["realtimeInput": ["audioStreamEnd": true]]))

        _ = await withCheckedContinuation { (continuation: CheckedContinuation<String?, Never>) in
            let resumeNow: String?? = lock.withLock {
                if turnDone || closed { return .some(finals) }
                waiter = continuation
                return .none
            }
            if let text = resumeNow { continuation.resume(returning: text) }
            DispatchQueue.global().asyncAfter(deadline: .now() + timeout) { [weak self] in
                self?.completeTurn()   // stalled: give up, the batch fallback takes over
            }
        }
        let text = normalResult()
        close()
        return text
    }

    private func normalResult() -> String? {
        lock.withLock { completedNormally && !finals.isBlank ? finals : nil }
    }

    public func close() {
        let task: URLSessionWebSocketTask? = lock.withLock {
            if closed { return nil }
            closed = true
            queued.removeAll()
            return self.task
        }
        task?.cancel(with: .normalClosure, reason: nil)
        completeTurn()
    }

    // MARK: messages

    func setupMessage() -> [String: Any] {
        var transcription: [String: Any] = [
            "mode": mode == .smart ? "SMART" : "VERBATIM",
            "languageCodes": languageCodes,
        ]
        if !vocabulary.isEmpty { transcription["customVocabulary"] = Array(vocabulary.prefix(1000)) }
        return ["setup": [
            "model": "models/\(Constants.liveModel)",
            "generationConfig": ["responseModalities": ["TEXT"]],
            "realtimeInputConfig": ["automaticActivityDetection": ["disabled": true]],
            "inputAudioTranscription": transcription,
        ]]
    }

    func audioMessage(_ pcm: Data) -> [String: Any] {
        ["realtimeInput": ["audio": ["data": pcm.base64EncodedString(),
                                     "mimeType": "audio/pcm;rate=\(Int(Constants.sampleRate))"]]]
    }

    // MARK: socket

    private func send(_ text: String) {
        guard let task = lock.withLock({ closed ? nil : self.task }) else { return }
        task.send(.string(text)) { [weak self] error in
            if error != nil { self?.fail(.connectionLost) }
        }
    }

    private func receive() {
        guard let task = lock.withLock({ closed ? nil : self.task }) else { return }
        task.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(.string(let s)): self.handle(Data(s.utf8)); self.receive()
            case .success(.data(let d)): self.handle(d); self.receive()
            case .success: self.receive()
            case .failure: self.fail(self.httpFailure(task) ?? .connectionLost)
            }
        }
    }

    private func httpFailure(_ task: URLSessionWebSocketTask) -> TranscriptionError? {
        guard let code = (task.response as? HTTPURLResponse)?.statusCode, code >= 300 else { return nil }
        switch code {
        case 400, 401, 403: return .invalidApiKey
        case 429: return .quotaExceeded
        default: return .server(code: code, detail: "")
        }
    }

    /// Handles one server message. Internal so tests can feed recorded messages.
    func handle(_ data: Data) {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else { return }

        if root["setupComplete"] != nil { becameReady() }

        if let error = root["error"] as? [String: Any] {
            let code = (error["code"] as? Int) ?? Int("\(error["code"] ?? "")") ?? 0
            let message = (error["message"] as? String) ?? ""
            fail(code == 429 || message.localizedCaseInsensitiveContains("quota") ? .quotaExceeded
                 : [400, 401, 403].contains(code) ? .invalidApiKey : .server(code: code, detail: message))
            return
        }

        guard let server = root["serverContent"] as? [String: Any] else { return }
        if let interim = (server["interimInputTranscription"] as? [String: Any])?["text"] as? String, !interim.isEmpty {
            onInterim?(interim)
        }
        if let chunk = (server["inputTranscription"] as? [String: Any])?["text"] as? String, !chunk.isEmpty {
            lock.withLock {
                if !finals.isEmpty && !finals.hasSuffix(" ") { finals += " " }
                finals += chunk.trimmingCharacters(in: .whitespaces)
            }
        }
        if root["generationComplete"] != nil || server["generationComplete"] != nil || server["turnComplete"] != nil {
            lock.withLock { if !closed { completedNormally = true } }
            completeTurn()
        }
    }

    /// On `setupComplete`: open the turn, then drain the queue. New audio keeps queueing until the
    /// queue is empty, so chunks always go out in the order they were recorded.
    private func becameReady() {
        let first: Bool = lock.withLock {
            if activityStarted || closed { return false }
            activityStarted = true
            return true
        }
        guard first else { return }
        send(Self.json(["realtimeInput": ["activityStart": [String: Any]()]]))
        while true {
            let batch: [Data] = lock.withLock {
                if closed { return [] }
                let q = queued
                queued.removeAll()
                if q.isEmpty { ready = true }
                return q
            }
            if batch.isEmpty { break }
            for chunk in batch { send(Self.json(audioMessage(chunk))) }
        }
    }

    private func fail(_ error: TranscriptionError) {
        let task: URLSessionWebSocketTask? = lock.withLock {
            if failure == nil { failure = error }
            if closed { return nil }
            closed = true
            return self.task
        }
        task?.cancel(with: .abnormalClosure, reason: nil)
        completeTurn()
    }

    private func completeTurn() {
        let (waiter, text): (CheckedContinuation<String?, Never>?, String) = lock.withLock {
            turnDone = true
            let w = self.waiter
            self.waiter = nil
            return (w, finals)
        }
        waiter?.resume(returning: text)
    }

    private static func json(_ object: [String: Any]) -> String {
        (try? JSONSerialization.data(withJSONObject: object)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    }
}

private extension String {
    var isBlank: Bool { trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
}
#endif

File: src/Fluent.Core/GeminiLiveClient.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json.Nodes;
using System.Threading.Channels;

namespace Fluent.Core;

/// <summary>Streams microphone audio to <c>gemini-3.5-transcribe-live</c> over the Live API WebSocket while
/// the user is still talking, so the transcript is (almost) ready the moment they stop. A port of the
/// Android and Mac <c>GeminiLiveClient</c>: same setup message, manual turn boundaries, same
/// finalise-then-await flow.
///
/// Audio sent before the socket is ready is queued and flushed on <c>setupComplete</c>. If the socket never
/// gets ready, or fails, <see cref="FinishAsync"/> returns null and the caller falls back to the batch
/// request with the full recording, so a turn is never lost.</summary>
public sealed class GeminiLiveClient : IDisposable
{
    public Uri Url { get; }
    readonly string apiKey;
    readonly TranscriptionMode mode;
    readonly IReadOnlyList<string> languageCodes;
    readonly IReadOnlyList<string> vocabulary;
    readonly TimeSpan connectTimeout;

    readonly object gate = new();
    readonly ClientWebSocket socket = new();
    readonly Channel<string> outbox = Channel.CreateUnbounded<string>(new UnboundedChannelOptions { SingleReader = true });
    readonly CancellationTokenSource life = new();
    bool ready, closed, activityStarted, turnDone, completedNormally;
    readonly List<byte[]> queued = [];
    readonly StringBuilder finals = new();
    TaskCompletionSource<bool>? waiter;

    public TranscriptionError? Failure { get; private set; }
    /// <summary>Latest interim (not yet final) text, for display.</summary>
    public Action<string>? OnInterim { get; set; }

    public GeminiLiveClient(string apiKey, TranscriptionMode mode, IReadOnlyList<string> languageCodes,
        IReadOnlyList<string> vocabulary, Uri? url = null, TimeSpan? connectTimeout = null)
    {
        this.apiKey = apiKey;
        this.mode = mode;
        this.languageCodes = languageCodes;
        this.vocabulary = vocabulary;
        Url = url ?? Constants.LiveUrl;
        this.connectTimeout = connectTimeout ?? Constants.LiveConnectTimeout;
        socket.Options.CollectHttpResponseDetails = true;
    }

    public bool IsReady { get { lock (gate) return ready && !closed; } }

    /// <summary>Opens the socket and sends the setup message. Returns immediately; audio can be sent at once.</summary>
    public void Connect()
    {
        var sep = string.IsNullOrEmpty(Url.Query) ? "?" : "&";
        var uri = new Uri(Url + sep + "key=" + Uri.EscapeDataString(apiKey));
        Enqueue(SetupMessage().ToJsonString());
        _ = Task.Run(async () =>
        {
            try
            {
                using var connectCts = CancellationTokenSource.CreateLinkedTokenSource(life.Token);
                connectCts.CancelAfter(connectTimeout);
                await socket.ConnectAsync(uri, connectCts.Token).ConfigureAwait(false);
            }
            catch
            {
                Fail(HttpFailure() ?? TranscriptionError.ConnectionLost);
                return;
            }
            _ = Task.Run(SendLoop);
            _ = Task.Run(ReceiveLoop);
        });
        // Give up on the socket if it is not ready in time; the batch fallback takes over.
        _ = Task.Delay(connectTimeout).ContinueWith(_ =>
        {
            bool isReady; lock (gate) isReady = ready;
            if (!isReady) Fail(TranscriptionError.ConnectionLost);
        }, TaskScheduler.Default);
    }

    /// <summary>Streams 16 kHz mono PCM16 samples. Safe to call from the audio thread.</summary>
    public void Send(ReadOnlySpan<short> pcm)
    {
        if (pcm.IsEmpty) return;
        var data = new byte[pcm.Length * 2];
        Buffer.BlockCopy(pcm.ToArray(), 0, data, 0, data.Length);   // little-endian on every Windows PC
        bool flushNow;
        lock (gate)
        {
            if (closed) return;
            if (!ready) { queued.Add(data); return; }
            flushNow = true;
        }
        if (flushNow) Enqueue(AudioMessage(data).ToJsonString());
    }

    /// <summary>Ends the utterance and waits for the final transcript. Null means "use the batch fallback".</summary>
    public async Task<string?> FinishAsync(TimeSpan? timeout = null)
    {
        bool isReady, started, alreadyDone;
        lock (gate) { isReady = ready && !closed; started = activityStarted; alreadyDone = turnDone; }
        if (alreadyDone) { Close(); return NormalResult(); }
        if (!isReady) { Close(); return null; }
        if (started) Enqueue(new JsonObject { ["realtimeInput"] = new JsonObject { ["activityEnd"] = new JsonObject() } }.ToJsonString());
        Enqueue(new JsonObject { ["realtimeInput"] = new JsonObject { ["audioStreamEnd"] = true } }.ToJsonString());

        Task wait;
        lock (gate)
        {
            if (turnDone || closed) wait = Task.CompletedTask;
            else { waiter = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously); wait = waiter.Task; }
        }
        // Stalled: give up, the batch fallback takes over.
        await Task.WhenAny(wait, Task.Delay(timeout ?? Constants.LiveFinalizeTimeout)).ConfigureAwait(false);
        CompleteTurn();
        var text = NormalResult();
        Close();
        return text;
    }

    string? NormalResult()
    {
        lock (gate) return completedNormally && !string.IsNullOrWhiteSpace(finals.ToString()) ? finals.ToString() : null;
    }

    public void Close()
    {
        lock (gate)
        {
            if (closed) { return; }
            closed = true;
            queued.Clear();
        }
        outbox.Writer.TryComplete();
        _ = CloseSocketAsync(WebSocketCloseStatus.NormalClosure);
        CompleteTurn();
    }

    public void Dispose() => Close();

    // messages

    internal JsonObject SetupMessage()
    {
        var transcription = new JsonObject
        {
            ["mode"] = mode.Raw(),
            ["languageCodes"] = new JsonArray(languageCodes.Select(c => (JsonNode)JsonValue.Create(c)!).ToArray()),
        };
        if (vocabulary.Count > 0)
            transcription["customVocabulary"] = new JsonArray(vocabulary.Take(1000).Select(v => (JsonNode)JsonValue.Create(v)!).ToArray());
        return new JsonObject
        {
            ["setup"] = new JsonObject
            {
                ["model"] = $"models/{Constants.LiveModel}",
                ["generationConfig"] = new JsonObject { ["responseModalities"] = new JsonArray("TEXT") },
                ["realtimeInputConfig"] = new JsonObject { ["automaticActivityDetection"] = new JsonObject { ["disabled"] = true } },
                ["inputAudioTranscription"] = transcription,
            },
        };
    }

    internal static JsonObject AudioMessage(byte[] pcm) => new()
    {
        ["realtimeInput"] = new JsonObject
        {
            ["audio"] = new JsonObject
            {
                ["data"] = Convert.ToBase64String(pcm),
                ["mimeType"] = $"audio/pcm;rate={Constants.SampleRate}",
            },
        },
    };

    // socket

    void Enqueue(string text)
    {
        lock (gate) if (closed) return;
        outbox.Writer.TryWrite(text);
    }

    async Task SendLoop()
    {
        try
        {
            await foreach (var text in outbox.Reader.ReadAllAsync(life.Token).ConfigureAwait(false))
            {
                if (socket.State != WebSocketState.Open) break;
                await socket.SendAsync(Encoding.UTF8.GetBytes(text), WebSocketMessageType.Text, true, life.Token).ConfigureAwait(false);
            }
        }
        catch (OperationCanceledException) { }
        catch { Fail(TranscriptionError.ConnectionLost); }
    }

    async Task ReceiveLoop()
    {
        var buffer = new byte[64 * 1024];
        using var message = new MemoryStream();
        try
        {
            while (socket.State == WebSocketState.Open)
            {
                var r = await socket.ReceiveAsync(buffer, life.Token).ConfigureAwait(false);
                if (r.MessageType == WebSocketMessageType.Close) break;
                message.Write(buffer, 0, r.Count);
                if (!r.EndOfMessage) continue;
                Handle(message.ToArray());
                message.SetLength(0);
            }
            // The server closed the socket. If the turn had not completed, the batch fallback takes over.
            Fail(CloseFailure() ?? TranscriptionError.ConnectionLost);
        }
        catch (OperationCanceledException) { }
        catch { Fail(TranscriptionError.ConnectionLost); }
    }

    TranscriptionError? HttpFailure()
    {
        var code = (int)socket.HttpStatusCode;
        if (code < 300) return null;
        return code switch
        {
            400 or 401 or 403 => TranscriptionError.InvalidApiKey,
            429 => TranscriptionError.QuotaExceeded,
            _ => TranscriptionError.Server(code, ""),
        };
    }

    TranscriptionError? CloseFailure()
    {
        var reason = socket.CloseStatusDescription ?? "";
        if (reason.Contains("API key", StringComparison.OrdinalIgnoreCase)) return TranscriptionError.InvalidApiKey;
        if (reason.Contains("quota", StringComparison.OrdinalIgnoreCase)) return TranscriptionError.QuotaExceeded;
        return null;
    }

    /// <summary>Handles one server message. Internal so tests can feed recorded messages.</summary>
    internal void Handle(byte[] data)
    {
        JsonObject? root;
        try { root = JsonNode.Parse(data) as JsonObject; } catch { return; }
        if (root is null) return;

        if (root.ContainsKey("setupComplete")) BecameReady();

        if (root["error"] is JsonObject error)
        {
            var code = error["code"] is JsonValue cv && cv.TryGetValue<int>(out var ci) ? ci
                : int.TryParse(GeminiClient.Str(error["code"]), out var cs) ? cs : 0;
            var message = GeminiClient.Str(error["message"]) ?? "";
            Fail(code == 429 || message.Contains("quota", StringComparison.OrdinalIgnoreCase) ? TranscriptionError.QuotaExceeded
                : code is 400 or 401 or 403 ? TranscriptionError.InvalidApiKey : TranscriptionError.Server(code, message));
            return;
        }

        if (root["serverContent"] is not JsonObject server) return;
        if (GeminiClient.Str((server["interimInputTranscription"] as JsonObject)?["text"]) is { Length: > 0 } interim)
            OnInterim?.Invoke(interim);
        if (GeminiClient.Str((server["inputTranscription"] as JsonObject)?["text"]) is { Length: > 0 } chunk)
        {
            lock (gate)
            {
                if (finals.Length > 0 && finals[^1] != ' ') finals.Append(' ');
                finals.Append(chunk.Trim(' ', '\t'));
            }
        }
        if (root.ContainsKey("generationComplete") || server.ContainsKey("generationComplete") || server.ContainsKey("turnComplete"))
        {
            lock (gate) if (!closed) completedNormally = true;
            CompleteTurn();
        }
    }

    /// <summary>On <c>setupComplete</c>: open the turn, then drain the queue. New audio keeps queueing until
    /// the queue is empty, so chunks always go out in the order they were recorded.</summary>
    void BecameReady()
    {
        lock (gate)
        {
            if (activityStarted || closed) return;
            activityStarted = true;
        }
        Enqueue(new JsonObject { ["realtimeInput"] = new JsonObject { ["activityStart"] = new JsonObject() } }.ToJsonString());
        while (true)
        {
            List<byte[]> batch;
            lock (gate)
            {
                if (closed) return;
                batch = [.. queued];
                queued.Clear();
                if (batch.Count == 0) ready = true;
            }
            if (batch.Count == 0) break;
            foreach (var chunk in batch) Enqueue(AudioMessage(chunk).ToJsonString());
        }
    }

    void Fail(TranscriptionError error)
    {
        lock (gate)
        {
            Failure ??= error;
            if (closed) { }
            else closed = true;
        }
        outbox.Writer.TryComplete();
        _ = CloseSocketAsync(WebSocketCloseStatus.EndpointUnavailable);
        CompleteTurn();
    }

    async Task CloseSocketAsync(WebSocketCloseStatus status)
    {
        try
        {
            if (socket.State == WebSocketState.Open)
            {
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
                await socket.CloseOutputAsync(status, null, cts.Token).ConfigureAwait(false);
            }
        }
        catch { }
        finally
        {
            life.Cancel();
            socket.Abort();
        }
    }

    void CompleteTurn()
    {
        TaskCompletionSource<bool>? w;
        lock (gate)
        {
            turnDone = true;
            w = waiter;
            waiter = null;
        }
        w?.TrySetResult(true);
    }
}
